### ISSUE TITLE
grpc.version is required in the properties to build GRPC apps

### ISSUE THREAD

**Issue by timyates** [micronaut-projects/micronaut-starter#1173]
In 3.4.x (pre-grpc bom), micronaut-core contained a managed version for grpc

https://github.com/micronaut-projects/micronaut-core/blob/0b91dbfc8410e70aa90e9aa670be959fad6167bf/gradle/libs.versions.toml#L44

This was then part of the [3.4.x bom](https://search.maven.org/artifact/io.micronaut/micronaut-bom/3.4.1/pom)

```
    <grpc.version>1.39.0</grpc.version>
```

When we build starter, we parse the BOM, and extract all of the properties out into a properties file [in this task](https://github.com/micronaut-projects/micronaut-starter/blob/3.4.x/buildSrc/src/main/groovy/io/micronaut/internal/starter/tasks/WriteMicronautVersionInfoTask.java#L66-L91)

So in 3.4.x, we get a file `starter-core/build/version-info/micronaut-versions.properties` which contains

```
grpc.version=1.39.0
```

This is used by our Gradle project builder to define the GRPC plugin version for the protobuf builder

For the Kotlin DSL:
https://github.com/micronaut-projects/micronaut-starter/blob/ec8d30a7521678d0341542aae3f6f869f32fa82c/starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/buildGradle.rocker.raw#L236-L240

And the Groovy DSL:
https://github.com/micronaut-projects/micronaut-starter/blob/ec8d30a7521678d0341542aae3f6f869f32fa82c/starter-core/src/main/java/io/micronaut/starter/feature/build/gradle/templates/buildGradle.rocker.raw#L253-L255

It is also used (silently) by the Maven `protoc-jar-maven-plugin` plugin the failure if this property isn't set [can be seen here](https://github.com/micronaut-projects/micronaut-starter/runs/5849279038?check_suite_focus=true)

## Solutions!

There are 2 solutions I can think of, and both of them don't taste 100% delicious.

I am also not sure if this is just an issue with the way we build starter (as I suspect), or whether it would be something that affects the wider community...

### 1.  Add the GRPC managed version back into the micronaut-core BOM

That is what I have done here with this PR: https://github.com/micronaut-projects/micronaut-core/pull/7190

#### The Good

This works.

#### The bad.

We have to keep the 2 in sync...  If we update the grpc version in the module, we MUST ensure the version is updated in the bom

### 2.  Change the property extraction task to pull out versions from boms defined in the parent core bom

This could be a global "extract all the versions from the entire micronaut bom tree", or a limited "we know we need grpc, so get the versions for just that bom as well"

#### The Good

This stops us having to keep things in sync.

#### The bad

We may have version collisions.  However, this could be minimized by limiting the boms we traverse to just those we know we need for the build.
We cannot prefix the version names with something to avoid collisions, as the Maven plugin requires __exactly__ `grpc.version`

**Comment by timyates** [comment]
@sdelamo PTAL, this issue is blocking [starter work](https://github.com/micronaut-projects/micronaut-starter/pull/1174) on 3.5.x

**Comment by alvarosanchez** [comment]
I have another alternative: use the Gradle API to lookup a dependency version transitively from the Micronaut BOM, e.g:

```java
VersionInfo.getBomVersion("io.grpc", "grpc-bom") // This would return the version of the `grpc-bom` artifact, which is coming transitively from `micronaut-grpc-bom`
```

**Comment by timyates** [comment]
We would still need it as a property for the maven plugin

**Comment by alvarosanchez** [comment]
Which Maven plugin? Ours?

**Comment by timyates** [comment]
No, the `protoc-jar-maven-plugin`

**Comment by alvarosanchez** [comment]
They don't require `grpc.version`. It's us: https://github.com/micronaut-projects/micronaut-core/blob/3.4.x/parent/build.gradle#L117

Looking at the plugin documentation, it's unfortunate that it needs a custom `<pluginArtifact>` tag to define a version, since that can't benefit from dependency management.

Considering that `micronaut-grpc-bom` already defines a `grpc.version` property, another option could be to change the `micronaut-bom` generation process to "inline / unroll" properties defined in any `io.micronaut.*:micronaut-*-bom`. But this may be a blasphemy to the Gradle gods, so summoning @melix here.

**Comment by timyates** [comment]
I wonder whether for now we just add the `managed-grpc-version` back into core ([as here](https://github.com/micronaut-projects/micronaut-core/pull/7190)) and create an issue to look for a better solution later

**Comment by sdelamo** [comment]
How about creating a Gradle task to parse the GRPC bom and generates a `starter-core/build/version-info/micronaut-grpc-versions.properties`.

**Comment by alvarosanchez** [comment]
> I wonder whether for now we just add the `managed-grpc-version` back into core ([as here](https://github.com/micronaut-projects/micronaut-core/pull/7190)) and create an issue to look for a better solution later

Given this is blocking working on 3.5.x, I'm fine with that.

**Comment by alvarosanchez** [comment]
> How about creating a Gradle task to parse the GRPC bom and generates a `starter-core/build/version-info/micronaut-grpc-versions.properties`.

@sdelamo GOTO https://github.com/micronaut-projects/micronaut-starter/issues/1173#issuecomment-1094671981

**Comment by timyates** [comment]
Ok, I think https://github.com/micronaut-projects/micronaut-starter/pull/1176 fixes this for starter...  Working on a fix for core

**Comment by timyates** [comment]
This should fix it for core https://github.com/micronaut-projects/micronaut-core/pull/7202

Hopefully

**Comment by timyates** [comment]
Fixed by #1176
